### ISSUE TITLE
STDP learning rule

### ISSUE THREAD

**Issue by jcmharry** [fangwei123456/spikingjelly#76]
Hi, the spikingjelly is really good and easy to use. One advice is do you consider integrating some Brain-like properties such as STDP in this library? Maybe it's a good way to add the advantage of spikingjelly.

**Comment by fangwei123456** [comment]
We have a very primary implement STDP: 

https://github.com/fangwei123456/spikingjelly/blob/4b9f796da343f8da2d2a5870c817b6cea60d0650/spikingjelly/clock_driven/layer.py#L872

We are trying to write a better version. But the progress is slow, because STDP is an unsupervised method and it is hard to evaluate our impelment is "correct". @Grasshlw Do you have interest in it?

**Comment by lvntky** [comment]
Hello even if I don't have much information about STDP, I want to contribute to this project, if you can help me a little or show the way I can start to implement SDTP @fangwei123456

**Comment by fangwei123456** [comment]
If the pre neuron fires a spike before the post neuron fires a spike, then the synaptic weight increases, and vice versa, if the post neuron fires a spike before the pre neuron fires a spike, the synaptic weight decreases. The biological experimental data are shown in the figure below, and the horizontal axis is the time difference between a pair of pulses released by pre neurons and post neurons, i.e. the vertical axis represents the percentage change in synaptic strength:t_{post}−t_{pre}
![image](https://user-images.githubusercontent.com/38352109/121455778-a21ea100-c9d7-11eb-9cda-43e1fe80ae40.png)

**Comment by fangwei123456** [comment]
This synaptic strength is related to the time of spikes and can be fitted using the following formula:
![image](https://user-images.githubusercontent.com/38352109/121456123-2113d980-c9d8-11eb-9730-373886dda914.png)

**Comment by fangwei123456** [comment]
According to "Morrison A, Diesmann M, Gerstner W. Phenomenological models of synaptic plasticity based on spiketiming[J]. Biological cybernetics, 2008, 98(6): 459-478.", we can use the trace, x_{j} for the pre neuron, and y_{i} for the post neuron, to implement STDP:
![image](https://user-images.githubusercontent.com/38352109/121456288-6afcbf80-c9d8-11eb-981c-048f0e4e781a.png)

where detla(t)=1 only when t=1, else delta(t) = 0. t_{i}^{f} is the firing time of a spike from pre neuron i, and t_{j}^{f} is the firing time of a spike from post neuron j. The weight is changed when a spike is fired by the pre or the post neuron:
![image](https://user-images.githubusercontent.com/38352109/121456533-d050b080-c9d8-11eb-9edc-3976e1300db1.png)

where F is a function to control weight's change.
![image](https://user-images.githubusercontent.com/38352109/121456569-de9ecc80-c9d8-11eb-9731-09235122217d.png)

**Comment by fangwei123456** [comment]
The following codes are implement by the trace method:
https://github.com/fangwei123456/spikingjelly/blob/4b9f796da343f8da2d2a5870c817b6cea60d0650/spikingjelly/clock_driven/layer.py#L872

Here is the example:
```
import torch
import torch.nn as nn
from spikingjelly.clock_driven import layer, neuron, functional
from matplotlib import pyplot as plt
import numpy as np
def f_pre(x):
    return x.abs() + 0.1

def f_post(x):
    return - f_pre(x)

fc = nn.Linear(1, 1, bias=False)

stdp_learner = layer.STDPLearner(100., 100., f_pre, f_post)
trace_pre = []
trace_post = []
w = []
T = 256
s_pre = torch.zeros([T, 1])
s_post = torch.zeros([T, 1])
s_pre[0: T // 2] = (torch.rand_like(s_pre[0: T // 2]) > 0.95).float()
s_post[0: T // 2] = (torch.rand_like(s_post[0: T // 2]) > 0.9).float()

s_pre[T // 2:] = (torch.rand_like(s_pre[T // 2:]) > 0.8).float()
s_post[T // 2:] = (torch.rand_like(s_post[T // 2:]) > 0.95).float()

for t in range(T):
    stdp_learner.stdp(s_pre[t], s_post[t], fc, 1e-2)
    trace_pre.append(stdp_learner.trace_pre.item())
    trace_post.append(stdp_learner.trace_post.item())
    w.append(fc.weight.item())

plt.style.use('science')
fig = plt.figure(figsize=(10, 6))
s_pre = s_pre[:, 0].numpy()
s_post = s_post[:, 0].numpy()
t = np.arange(0, T)
plt.subplot(5, 1, 1)
plt.eventplot((t * s_pre)[s_pre == 1.], lineoffsets=0, colors='r')
plt.yticks([])
plt.ylabel('$S_{pre}$', rotation=0, labelpad=10)
plt.xticks([])
plt.xlim(0, T)
plt.subplot(5, 1, 2)
plt.plot(t, trace_pre)
plt.ylabel('$tr_{pre}$', rotation=0, labelpad=10)
plt.xticks([])
plt.xlim(0, T)

plt.subplot(5, 1, 3)
plt.eventplot((t * s_post)[s_post == 1.], lineoffsets=0, colors='r')
plt.yticks([])
plt.ylabel('$S_{post}$', rotation=0, labelpad=10)
plt.xticks([])
plt.xlim(0, T)
plt.subplot(5, 1, 4)
plt.plot(t, trace_post)
plt.ylabel('$tr_{post}$', rotation=0, labelpad=10)
plt.xticks([])
plt.xlim(0, T)
plt.subplot(5, 1, 5)
plt.plot(t, w)
plt.ylabel('$w$', rotation=0, labelpad=10)
plt.xlim(0, T)

plt.show()
```

![image](https://user-images.githubusercontent.com/38352109/121456918-67b60380-c9d9-11eb-874b-f4b0d932570b.png)

**Comment by fangwei123456** [comment]
#15

**Comment by lvntky** [comment]
okay @fangwei123456 thank you very much for this great reference, I forked the repo and started to improve as I can, I will open a PR if I can make anything valuable. Also is there any discord channel, or email that I can talk with you about the progress of code or should I just write here?

**Comment by fangwei123456** [comment]
@lvntky Thanks! I suggest that you can just write here, which is convenient for other developers to join!

**Comment by fangwei123456** [comment]
@lvntky @jcmharry 
Hi, STDP is provided now:

https://spikingjelly.readthedocs.io/zh_CN/latest/activation_based_en/stdp.html
